azkaban-execserver/src/main/java/azkaban/execapp/StatisticsServlet.java

```diff
@@ -0,0 +1,287 @@
+              long tmp_freeMemory  = 0 ;
+              try {
+                tmp_totalMemory = Long.parseLong(splitedresult[1]);
+                tmp_freeMemory = Long.parseLong(splitedresult[3]);
+
+              } catch(NumberFormatException e){
+                logger.error("skipping the unprocessable line from the output -" + line);
+                continue;
+              }
+
+              // add up the result.
+              ++sampleCount;
+              totalMemory += tmp_totalMemory ;
+              freeMemory  += tmp_freeMemory;
+            }
+          }
+
+          // set the value.
+          if (sampleCount > 0){
+            freeMemory  = freeMemory  / sampleCount;
+            totalMemory = totalMemory / sampleCount;
+            logger.info(String.format("total memory - %s , free memory - %s", totalMemory, freeMemory));
+            stats.setRemainingMemory(freeMemory);
+            stats.setRemainingMemoryPercent(totalMemory == 0? 0 :
+              ((double)freeMemory/(double)totalMemory));
+          }
+        }
+      }
+      catch (Exception ex){
+        logger.error("failed fetch system memory info " +
+                     "as exception is captured when fetching result from bash call.");
+      }
+    } else {
+        logger.error("failed fetch system memory info " +
+                     "as 'bash' or 'free' command can't be found on the current system.");
+    }
+  }
+
+  /**
+   * call the data providers to fill the returning data container for statistics data.
+   * This function refreshes the static cached copy of data in case if necessary.
+   * */
+  protected synchronized void populateStatistics(boolean noCache){
+    //check again before starting the work.
+    if (noCache || null == lastRefreshedTime ||
+        new Date().getTime() - lastRefreshedTime.getTime() > cacheTimeInMilliseconds){
+      final Statistics stats = new Statistics();
+
+      List<Thread> workerPool = new ArrayList<Thread>();
+      workerPool.add(new Thread(new Runnable(){ public void run() {
+        fillRemainingMemoryPercent(stats); }},"RemainingMemoryPercent"));
+
+      workerPool.add(new Thread(new Runnable(){ public void run() {
+        fillRemainingFlowCapacityAndLastDispatchedTime(stats); }},"RemainingFlowCapacityAndLastDispatchedTime"));
+
+      workerPool.add(new Thread(new Runnable(){ public void run() {
+        fillCpuUsage(stats); }},"CpuUsage"));
+
+      // start all the working threads.
+      for (Thread thread : workerPool){thread.start();}
+
+      // wait for all the threads to finish their work.
+      // NOTE: the result container itself is not thread safe, we are good as for now no
+      //       working thread will modify the same property, nor have any of them
+      //       need to compute values based on value(s) of other properties.
+      for (Thread thread : workerPool){
+        try {
+          // we gave maxim 5 seconds to let the thread finish work.
+          thread.join(5000);;
```

Comment:
function is synchronized so it is guarded, and please see the logic in "doGet", the whole mechanism  is implemented in the classic "double checked locking" manner to ensure thread safety.